### ISSUE TITLE
Plan: Modernizing F# Analysis

### ISSUE THREAD

**Issue by TIHan** [dotnet/fsharp#11976]
Don and I sat down to document my overall plan for modernizing the way F# and FCS do analysis.  Much of this work is already done, this documents the plan end-to-end.  Here's what we came up with, please comment and discuss below.  

# Planning: Modernizing F# Analysis

This note describes the technical agenda to "modernize" the FCS analysis services to use best-known techniques from Roslyn.  

In the long term this agenda delivers multiple critical benefits:

1. It implements high-performance multi-threaded analysis

2. It gives a more reliable basis for implementing multiple IDE features, including cross-file refactorings and analysis 

3. It implements "in-memory documents" and "in-memory cross-project references to C#", simplifying the user experience of using F# in Visual Studio.

4. It a strong basis for reliably make F# analysis more incremental w.r.t. incremental changes in inputs.

5. It a strong basis for reliably building an "out-of-proc" LSP implementation for F# following Roslyn design principles.

6. It aligns F# with the architectural principles of Roslyn, allowing contributors to transfer experience between the two

The core of the plan is to adopt a more Roslyn-like model of analysis, based on 

1. Immutable snapshots of the contents of documents and projects

2. Immutable views of their enrichment with analysis information

3. A cacheless compiler-service API

### The Current Situation and why it's a Problem

FCS provides services to compute analysis information from inputs. 
For the current API, some of these inputs are filenames, and so FCS relies in part on the state of the file system, which is highly mutable state and is highly problematic.  

Specifically, when requesting the analysis of a file in a project (e.g. for a refactoring or a tooltip), the state of the **current** file is captured as a "snapshot", but the state of **other files** in the project is accessed via the file system as the analysis proceeds.  This causes four problems:

A. The state of these files as saved on disk may have changed in-between
B. Differences between the saved and unsaved contents of in-memory buffers in the IDE.  
C. It is extremely error-prone to implement incremental updates to analysis w.r.t. incremental changes in input

Additionally, FCS had two other major problems:
D. FCS is stateful and implements multiple kinds of caching for parsing and analysis
E. FCS was single-threaded, with a "reactor thread" compilation lock

Problem A leads to:

1. Repeated polling checks on timestamps of files whenever checking for validity of the results 

2. Is a very frequent cause of bugs (BUG LINKS)

Problem B leads to:

1. Cognitive overload for the user who doesn't understand that prior files must be saved. That is, F# doesn't support 

2. Double type-checking of open files in a project when a change is made: once as part of the "background" build that is done using on-disk representations, once as part of the "foreground" build in order to get diagnostics for currently open documents.  This is not something users perceive, but is a potential overall performance gain we can deliver for large projects.  

3. Unnecessary complexity and distinctions in the FCS API that can makes it difficult to understand what's going - for example, we must document and test the differences between foreground and background checking. 

4. A slew of bugs in cross-file refactorings (LINK LINK), cross-file goto-definition, cross-file tooltips (based on saved, not unsaved contents)

5. Missed opportunities to take advantage of F# language features for more efficient incremental checking, in particular signature files.

Problems C & D leads to a slew of bugs related to not invalidating cache entries with regard to changes in on-disk files.  Problem D also causes issues with memory usage and too many analysis results being "kept live" by the FCS caches.

Problems A-C also apply to the "referenced assemblies" inputs to analysis, particularly cross-project references.
Before the start of this work, specifying a cross-project reference was done via a graph of FSharpProjectOptions,
but no in-memory cross-project references were allowed to C# projects. Further, the cross-project references
lead to reading input files from disk for other projects and assessing their timestamps, leading to bugs
and inconsistencies.

In combination these issues lead to a kind of "grid lock" where the root causes of the kinds of bugs we see are not addressed using best-known techniques.  We patch a few bugs, which can cause other bugs etc.   We know the solution to unlock this, which is to follow the design principles used by Roslyn.

> Aside: Problem B can be partly addressed by an existing "hack" in the FCS API that allows the file system used by FCS to be "shimmed".  This is used by JetBrains Rider in order to implement in-memory documents.  However this is an awkward solution that differs greatly from the Roslyn approach, and Problems A, C and D still remains.

### What's Needed

The Roslyn approach to these problems is to 

1. Make all inputs to analysis be "snapshot" objects

2. Make all analysis results to be on-demand stateless enrichments of these snapshots

3. Do not implement adhoc caching of analysis objects within Roslyn, but rather allow liveness of analysis objects to determine lifetimes. 

The technical agenda is based on transforming FCS to correspond to these principles.

> Aside: when we say Roslyn analysis objects (e.g. `Compilation` LINK) are "on-demand stateless enrichments", this means there may be internal state recording what enrichments have already been computed, and this may be important or reasoning about memory usage.  However, logically speaking, the analysis objects are still functional enrichments.  Roslyn analysis objects are effectively "incrementally computed".

### Technical Agenda

The agenda is as follows.  Where new constructs are brought into existence in the FCS API, we show their correspondence to Roslyn equivalents

* [x] [Add IFSharpSourceText](https://github.com/dotnet/fsharp/pull/6001) (corresponds to [SourceText](https://docs.microsoft.com/en-us/dotnet/api/microsoft.codeanalysis.text.sourcetext?view=roslyn-dotnet-3.11.0) in Roslyn).  This allows for immutable views of snapshots of buffers

      This included adopting IFSharpSourceText for foreground analysis. 
      
      OBSERVABLE GAIN: Among other things this prevented copying entire source files, a major cause of GC,
      and removed a slew of bugs and workarounds related to out-of-date snapshots.

* [x] [Rewrite the "IncrementalBuild" engine to use a build graph](https://github.com/dotnet/fsharp/pull/11573)

      This allows for simple and reliable implementation of on-demand stateless enrichments within the background build.

      There were also several major cleanup steps preparing for this.  For example
      the build graph must be correctly incremental between the "diagnostics+tooltips" portion of
      analysis results and the more costly "symbol usages" portion, see https://github.com/dotnet/fsharp/pull/11666.

      OBSERVABLE GAIN: This enabled "in-memory cross project referencing" for C# -> F# projects.  This gives a much
      simpler user experience, because changes in C# projects are now reflected immediately in a F# project
      without having to compile the C# project on-disk, and analysis results are available in uncompiled solutions.
            
* [x] [Make the build graph free-threaded](https://github.com/dotnet/fsharp/pull/11573)

      OBSERVABLE GAIN: With this change, FCS started supporting concurrent requests for analysis results.  This massively improves
      the performance of analysis and the responsiveness of the IDE. 

* [x] [Make the build graph more incremental w.r.t. changes in implementation file without change in signature file](https://github.com/dotnet/fsharp/pull/10199)

      OBSERVABLE GAIN: This greatly improved performance of analysis in F# projects that contain signature files. The user
      sees diagnostics and other analysis results much more quickly when a change is made to an implementation file without
      change to the signature file.

* [x] [Add FSharpReferencedProject](https://github.com/dotnet/fsharp/pull/11301), corresponding to [Roslyn's CompilationReference](https://docs.microsoft.com/ko-kr/dotnet/api/microsoft.codeanalysis.compilationreference?view=roslyn-dotnet)

      OBSERVABLE GAIN: This enabled "in-memory cross project referencing" for C# -> F# projects.  This gives a much
      simpler user experience, because changes in C# projects are now reflected immediately in a F# project
      without having to compile the C# project on-disk, and analysis results are available in uncompiled solutions.

* [ ] Add FSharpSource, corresponding to Roslyn's SyntaxTree.  A prototype is in [this PR](https://github.com/dotnet/fsharp/pull/11588). 
      
      An FSharpSource is an input to analysis and internally is an "IFSharpSourceText + F# parse tree". The background and foreground analysis routines will accept these objects.

      NOTE: These are not incremental w.r.t. incremental changes in the IFSharpSourceText (adding, removing lines)
      though in theory they could allow some incrementality in future refinements. 

* [ ] Add FSharpProject, corresponding to Roslyn's Compilation. A prototype is in [this PR](https://github.com/dotnet/fsharp/pull/11775). 

      An FSharpProject is a handle to the "outputs" of analysis. That is, an FSharpProject is a Roslyn-like on-demand analysis object
      which can be used to request analysis information, e.g. diagnostics, tool tips, symbols uses, across an entire project.

      An FSharpProject is incremental w.r.t. to replacing the FSharpSource inputs. That is, the incrementality granularity
      is "replace the contents of an entire file". If, for example, the last FSharpSource in the compilation sequence is
      replaced, then the majority of the results of analysis will be re-used.

      An FSharpProject is on-demand in multiple ways, all mediated by the internal build graph.  For example, if
      diagnostics are requested for the 3rd file out of 5 in the compilation sequence, only 3 files will be checked. Id
      the diagnostics are then requested for the 5th file, the remaining two files will be checked.  Some semantic
      information may be re-computed each time it is requested.  Some may require more detail re-checking, e.g. to record
      symbol locations.

      An FSharpProject can be created without needing an FSharpChecker. Thus they are not tied to the caches of FSharpChecker.
      
      This addresses problem (D) above, in the sense that FCS clients like Visual Studio can choose their own
      lifetimes for FSharpProject objects (usually the lifetime of the associated Roslyn Workspace).
      
* [ ] Have Visual Studio provide FSharpSource objects based on live buffers.
      A prototype is part of https://github.com/dotnet/fsharp/pull/11588. 
 
      OBSERVABLE GAIN: This reliably implements the "in-memory documents" feature

      OBSERVABLE GAIN: This gives reliable rename-refactor for unsaved files.

      OBSERVABLE GAIN: This avoids a slew of other bugs and complexity going forward
      
* [ ] Have Visual Studio use FSharpProject objects instead of FSharpChecker.
      A prototype is part of https://github.com/dotnet/fsharp/pull/11775. 
      A preparatory PR was https://github.com/dotnet/fsharp/pull/11694

      OBSERVABLE GAIN: This reduces memory usage
      
      OBSERVABLE GAIN: This reduces bugginess due to invalidation, timing and state problems
      
      This is also preparatory to reliably making FCS out-of-process with LSP.

* [ ] Stabilize, document, refine the public APIs as part of making FCS a binary compatible component in support of [F# Analyzers](https://github.com/dotnet/fsharp/pull/11057)

#### Further Incrementality

One result of the above agenda is that is provides a basis to begin to implement finer-grained incremental
adjustment of analysis results w.r.t. incremental changes in inputs.   Currently (at the end of the above
agenda) incrementality is at the granularity of replacing the contents of an entire file.  We could
now consider incrementality w.r.t. adding text at the end of a file, or changes within a line.  This
requires incremental parsing, checking.  The aim here would be higher performance IDE analysis.

Roslyn supports this kind of incrementality but it is not an essential part of the above agenda.

#### LSP and Out of Process

Future changes to Roslyn will require F# to implement LSP, doing diagnostic analysis out-of-process (see https://github.com/dotnet/fsharp/issues/11969). 

An LSP implementation of F# will host FSharp.Compiler.Service and should ideally have an implementation
architecture very similar to the C# out-of-proc LSP implementation.  Completing this agenda allows
us to use this approach.  For example, the out-of-proc process will mirror the Roslyn workspace and
hold handles to the appropriate FSharpProject objects, just as the C# version of the same
holds a Roslyn Compilation object.

Crucially, this means the F# LSP implementation will be simple, reliable and relatively stateless (apart from holding FSharpProject
objects).

**Comment by alfonsogarciacaro** [comment]
This is great work @TIHan, looking forward to it!

Not entirely sure if it's related but just to mention that it'd be nice if this work also takes into account FSC-based compilers targeting other platforms. Particularly about incremental compilation. For Fable, we use a [custom build of FSC from a branch of @ncave fork](https://github.com/ncave/fsharp/blob/d3fd0b7ebc9e3389ab1349bb3d4979e8da652697/src/fsharp/FSharp.Compiler.Service/service_slim.fs) that does some [simple caching](https://github.com/ncave/fsharp/blob/d3fd0b7ebc9e3389ab1349bb3d4979e8da652697/src/fsharp/FSharp.Compiler.Service/service_slim.fs#L159-L161) (it only recompiles the first changed file and those below) and skips work that's not needed to get the typed AST (basically symbol information, if I'm not mistaken). It'd be great if we could have a similar mechanism directly built into FSC.

**Comment by kerams** [comment]
I assume out of process hosting implies leaving the .NET Framework world, with free runtime performance gains, type providers' design time parts being able to target .NET 6, etc.? Can't wait.

Any updates on how it's going and what the plans for the immediate future are? Obviously Will's leaving the team has thrown a wrench into this endeavor a bit.
